lib/xml_security.rb

```diff
@@ -161,10 +161,10 @@ def sign_document(private_key, certificate, signature_method = RSA_SHA1, digest_
       # add the signature
       issuer_element = self.elements["//saml:Issuer"]
       if issuer_element
-        self.root.insert_after issuer_element, signature_element
+        self.root.insert_after(issuer_element, signature_element)
       else
-        if sp_sso_descriptor = self.elements["/md:EntityDescriptor"]
-          self.root.insert_before sp_sso_descriptor, signature_element
+        if sp_sso_descriptor = self.elements["/md:EntityDescriptor/md:SPSSODescriptor"]
+          self.root.insert_before(sp_sso_descriptor, signature_element)
```

Comment:
@pitbulk the code as I have it here is correct with what you are saying. `self.root` is `EntityDescriptor`, so `self.root.insert_before(sp_sso_descriptor, ...)` means "Insert with `EntityDescriptor` as parent, as a sibling immediately before `md:SPSSODescriptor`)

Reply:
Ok, in case of md:EntitiesDescriptor Signature needs to be added as first child which is covered later... I think you are right.... maybe we could add a unittest to verify where the signature is added on the different scenarios to assure it is not break in the future. Rather than that the PR seems ok, I just had a minus comment about a refactor. Good job.

Reply:
In that case we should force it to always insert explicitly as the first child of `md:EntitiesDescriptor`. The way the code is now, it "happens to work" because `md:SPSSODescriptor` happens to be the first child. But if one's subclass were to do some crazypants custom modification of the xml in the `add_extras` method, this might no longer be the case.

Reply:
I'm not an expert in REXML so I'll have to check the right way to do this. If you know please let me know :)

Reply:
Not an REXML expert neither ;)

Reply:
Unit test added and ready for final review. Note that unit test includes `validate_xml!` which catches most of errors. In a separate PR we should consider to have the Metadata class itself do the XSD validation, since we're now allowing users to add custom elements and the validation is quite strict.